### PR TITLE
docs: detail how to mount and use external drivers

### PR DESCRIPTION
**Pull request by adwk67**
# Description

docs: detail how to mount and use external drivers.
fixes https://github.com/stackabletech/hive-operator/issues/416.

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes
 
```[tasklist]
# Author
- [x] Changes are OpenShift compatible
- [x] ~Integration tests passed (for non trivial changes)~ Tested with integration test in local branch
```

```[tasklist]
# Reviewer
- [ ] Code contains useful comments
- [ ] Code contains useful logging statements
- [ ] (Integration-)Test cases added
- [x] Documentation added or updated. Follows the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs-style-guide).
- [x] Changelog updated
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
- [ ] [Roadmap](https://github.com/orgs/stackabletech/projects/25/views/1) has been updated
```

### CODE REVIEW

File: docs/modules/hive/pages/usage-guide/database-driver-example.adoc [hunk 1]
```diff
@@ -0,0 +1,176 @@
+
+= Database drivers
+
+The Stackable Apache Hive product images come with built-in support for using Postgresql as the metastore  database. To use another supported database it is necessary to make the relevant drivers available to Hive: this tutorial shows how this is done for MySql.
```
Comment (maltesander) [inline docs/modules/hive/pages/usage-guide/database-driver-example.adoc:4]:
```suggestion
The Stackable Apache Hive product images come with built-in support for using Postgresql as the metastore database. To use another supported database it is necessary to make the relevant drivers available to Hive: this tutorial shows how this is done for MySql.
```

Comment (maltesander) [inline docs/modules/hive/pages/usage-guide/database-driver-example.adoc:4]: Official names use capitalized SQL: PostgreSQL, MySQL

File: docs/modules/hive/pages/usage-guide/database-driver-example.adoc [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+  name: hive-with-drivers
+spec:
+  image:
+    productVersion: 3.1.3
+  clusterConfig:
+    database:
+      connString: jdbc:mysql://mysql:3306/hive  # <1>
+      user: hive  # <2>
+      password: hive
+      dbType: mysql
```
Comment (maltesander) [inline docs/modules/hive/pages/usage-guide/database-driver-example.adoc:147]: We should not forget to update this when moving the credentials to a secret.
Edit: You mentiond it in the footnotes below, do we want to add this to the credentials ticket though?

File: docs/modules/hive/pages/usage-guide/database-driver-example.adoc [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    s3:
+      reference: minio  # <3>
+  metastore:
+    roleGroups:
+      default:
+        envOverrides:
+          METASTORE_AUX_JARS_PATH: "/stackable/external-drivers/mysql-connector-j-8.0.31.jar"   # <4>
```
Comment (maltesander) [inline docs/modules/hive/pages/usage-guide/database-driver-example.adoc:154]: Did not test this, would prefer this if multiple drivers are added?
```suggestion
          METASTORE_AUX_JARS_PATH: "/stackable/external-drivers/*"   # <4>
```

**Review by maltesander — CHANGES_REQUESTED** [review]
Just minor things, LGTM otherwise.

File: docs/modules/hive/pages/usage-guide/database-driver-example.adoc (hunk 2 above)
Comment (adwk67) [reply docs/modules/hive/pages/usage-guide/database-driver-example.adoc:147]: I think it is already covered there with this comment:

>  integration tests, examples and documentation have been updated

File: docs/modules/hive/pages/usage-guide/database-driver-example.adoc (hunk 1 above)
Comment (adwk67) [reply docs/modules/hive/pages/usage-guide/database-driver-example.adoc]: Thanks for spotting that! Done.

**Comment by adwk67** [comment]
@maltesander I have a local integration test for this. What do you think - shall I push that too and just leave it commented out of the test definitions (so it can be activated as necessary)? I wouldn't leave it active as we would either have to upload the mysql driver to Nexus or have an external dependency.

**Comment by fhennig** [comment]
I wanna look at it more in depth, but already I'd like the fileame to just be database-driver.adoc and also if you could put linebreaks between sentences (so each sentence is on its own line) that would be great!

File: docs/modules/hive/pages/usage-guide/database-driver-example.adoc (hunk 3 above)
Comment (maltesander) [reply docs/modules/hive/pages/usage-guide/database-driver-example.adoc:154]: I saw some examples with the comma seperated list, but people said pointing e.g. to hdfs://.../* is fine. So i would assume it works but fine with leaving it as is.

**Comment by maltesander** [comment]
> @maltesander I have a local integration test for this. What do you think - shall I push that too and just leave it commented out of the test definitions (so it can be activated as necessary)? I wouldn't leave it active as we would either have to upload the mysql driver to Nexus or have an external dependency.

I dont think its in scope here, but having an integration test (could be an existing one) switching to mysql for testing would be nice :)

File: docs/modules/hive/pages/required-external-components.adoc [hunk 4]
```diff
@@ -8,3 +8,5 @@ The Hive Metastore requires a backend SQL database. Supported databases and vers
 * MS SQL Server 2008 R2 and above
 
 Reference: https://cwiki.apache.org/confluence/display/Hive/AdminManual+Metastore+Administration#AdminManualMetastoreAdministration-SupportedBackendDatabasesforMetastore[Hive Metastore documentation]
+
+The Stackable product images for Apache Hive come with built-in support for Postgresql. See xref:usage-guide/database-driver-example.adoc[] for details on how to make drivers for other databases (supported by Hive) available.
```
Comment (fhennig) [inline docs/modules/hive/pages/required-external-components.adoc:12]:
```suggestion
The Stackable product images for Apache Hive come with built-in support for PostgreSQL.
See xref:usage-guide/database-driver.adoc[] for details on how to make drivers for other databases (supported by Hive) available.
```

File: docs/modules/hive/pages/usage-guide/database-driver.adoc [hunk 5]
```diff
@@ -0,0 +1,182 @@
+
+= Database drivers
+
+The Stackable Apache Hive product images come with built-in support for using PostgreSQL as the metastore  database.
+
+To use another supported database it is necessary to make the relevant drivers available to Hive: this tutorial shows how this is done for MySQL.
```
Comment (fhennig) [inline docs/modules/hive/pages/usage-guide/database-driver.adoc:6]:
```suggestion
The Stackable product images for Apache Hive come with built-in support for using PostgreSQL as the metastore  database.
To use another supported database it is necessary to make the relevant drivers available to Hive: this tutorial shows how this is done for MySQL.
```

File: docs/modules/hive/pages/usage-guide/database-driver.adoc [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+
+[source,bash]
+----
+helm install mysql oci://registry-1.docker.io/bitnamicharts/mysql \
+--set auth.database=hive \
+--set auth.username=hive \
+--set auth.password=hive
+----
+
+== Download the driver to a `PersistentVolumeClaim`
```
Comment (fhennig) [inline docs/modules/hive/pages/usage-guide/database-driver.adoc:18]:
```suggestion
== Download the driver to a PersistentVolumeClaim
```

File: docs/modules/hive/pages/usage-guide/database-driver.adoc [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+              mountPath: /stackable/externals
+----
+
+== Create a Hive Cluster
+
+The MySQL connection details can then be used in the definition of the Hive Metastore resource.
+
+Note that it is also necessary to "tell" Hive where to find the driver.
+
+This is done by setting the value of the environment variable `METASTORE_AUX_JARS_PATH` to the path of the mounted driver:
```
Comment (fhennig) [inline docs/modules/hive/pages/usage-guide/database-driver.adoc:136]:
```suggestion
The MySQL connection details can then be used in the definition of the Hive Metastore resource.
Note that it is also necessary to "tell" Hive where to find the driver.
This is done by setting the value of the environment variable `METASTORE_AUX_JARS_PATH` to the path of the mounted driver:
```

File: docs/modules/hive/pages/usage-guide/database-driver.adoc [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+                  - name: external-drivers
+                    mountPath: /stackable/external-drivers
+            volumes:
+              - name: external-drivers
+                persistentVolumeClaim:
+                  claimName: pvc-hive-drivers
+        replicas: 1
+----
+
+<1> The database connection details matching those given when deploying the MySQL helm chart
```
Comment (fhennig) [inline docs/modules/hive/pages/usage-guide/database-driver.adoc:175]:
```suggestion
<1> The database connection details matching those given when deploying the MySQL Helm chart
```

File: docs/modules/hive/pages/usage-guide/database-driver.adoc [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+<2> Plain-text Hive credentials will be replaced in an upcoming release!
+<3> A reference to the file store using S3 (this has been omitted from this article for the sake of brevity, but is described in e.g. the xref:getting_started/first_steps.adoc[] guide)
+<4> Use `envOverrides` to set the driver path
+<5> Use `podOverrides` to mount the driver
+
+[sidebar]
+This has been tested on Azure AKS and Amazon EKS, both running Kubernetes 1.29. The example shows a PVC mounted with the access mode `ReadWriteOnce` as we have a single metastore instance that is deployed only once the jobs have completed, and, so long as these all run after each other, they can be deployed to different nodes. Different scenarios may require a different access mode, the availability of which is dependent on the type of cluster in use.
```
Comment (fhennig) [inline docs/modules/hive/pages/usage-guide/database-driver.adoc:182]:
```suggestion
NOTE: This has been tested on Azure AKS and Amazon EKS, both running Kubernetes 1.29.
The example shows a PVC mounted with the access mode `ReadWriteOnce` as we have a single metastore instance that is deployed only once the jobs have completed, and, so long as these all run after each other, they can be deployed to different nodes.
Different scenarios may require a different access mode, the availability of which is dependent on the type of cluster in use.
```

File: docs/modules/hive/pages/usage-guide/database-driver.adoc [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+              "pipefail",
+              "-c",
+              "stat ${DEST_DIR}/mysql-connector-j-8.0.31.jar",
+            ]
+          volumeMounts:
+            - name: external-drivers
+              mountPath: /stackable/externals
+----
+
+== Create a Hive Cluster
```
Comment (fhennig) [inline docs/modules/hive/pages/usage-guide/database-driver.adoc:130]:
```suggestion
== Create a Hive cluster
```

File: docs/modules/hive/pages/usage-guide/database-driver.adoc [hunk 11]
```diff
@@ -0,0 +1,182 @@
+
+= Database drivers
```
Comment (fhennig) [inline docs/modules/hive/pages/usage-guide/database-driver.adoc:2]:
```suggestion
= Database drivers
```

**Review by fhennig — APPROVED** [review]
Just a few things, thanks for documenting this!

File: docs/modules/hive/pages/required-external-components.adoc (hunk 4 above)
Comment (adwk67) [reply docs/modules/hive/pages/required-external-components.adoc]: Thanks for catching that!

File: docs/modules/hive/pages/usage-guide/database-driver.adoc [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+
+The Stackable product images for Apache Hive come with built-in support for using PostgreSQL as the metastore  database.
```
Comment (sbernauer) [inline docs/modules/hive/pages/usage-guide/database-driver.adoc:3]:
```suggestion
The Stackable product images for Apache Hive come with built-in support for using PostgreSQL as the metastore database.
The MySQL driver is not shipped in our images due to licensing problems.
```

**Review by sbernauer — CHANGES_REQUESTED** [review]
Thanks for documenting this!
Would you be willing to also document a small Dockerfile that does something like
```
FROM <our image>
curl -o /stackable/external-drivers/mysql-connector-j-8.0.31.jar" https://repo1.maven.org/maven2/com/mysql/mysql-connector-j/8.0.31/mysql-connector-j-8.0.31.jar
```

This should be very easy and help a lot for a lot of customers (that don't have or want ReadWriteMany). IMHO we should also recommend this and only have the PVC part as a last resort but that's just my 2 cents :see_no_evil:

**Comment by fhennig** [comment]
I think it's a good idea to document the alterantives as well, but it wasn't part of the original scope ofthe ticket. But ideally we'd lay out the options for the users, and then they can decide (Do I want to host an image somewhere? Can I use ReadWriteMany? etc)

**Comment by sbernauer** [comment]
I also volunteer to quickly throw the Dockerfile together, I don't want to put the burden on @adwk67. Shall I do that?

**Review by sbernauer — APPROVED** [review]
I was just able to had a look at this and you pushed the solution.
Many thanks for adding the Dockerfile! It builds on my machine and I assume you tested if the image works against mysql.

**Pull request merged**